lib/base/io-engine.cpp

```diff
@@ -16,60 +16,73 @@
 
 using namespace icinga;
 
-CpuBoundWork::CpuBoundWork(boost::asio::yield_context yc)
+CpuBoundWork::CpuBoundWork(boost::asio::yield_context yc, boost::asio::io_context::strand& strand)
 	: m_Done(false)
 {
 	auto& ioEngine (IoEngine::Get());
+	auto& sem (ioEngine.m_CpuBoundSemaphore);
+	std::unique_lock<std::mutex> lock (sem.Mutex);
 
-	for (;;) {
-		auto availableSlots (ioEngine.m_CpuBoundSemaphore.fetch_sub(1));
+	if (sem.FreeSlots) {
+		--sem.FreeSlots;
+		return;
+	}
+
+	auto cv (Shared<AsioConditionVariable>::Make(ioEngine.GetIoContext()));
+	bool gotSlot = false;
+	auto pos (sem.Waiting.insert(sem.Waiting.end(), IoEngine::CpuBoundQueueItem{&strand, cv, &gotSlot}));
 
-		if (availableSlots < 1) {
-			ioEngine.m_CpuBoundSemaphore.fetch_add(1);
-			IoEngine::YieldCurrentCoroutine(yc);
-			continue;
+	lock.unlock();
+
+	try {
+		cv->Wait(yc);
+	} catch (...) {
+		std::unique_lock<std::mutex> lock (sem.Mutex);
+
+		if (gotSlot) {
+			lock.unlock();
+			Done();
+		} else {
+			sem.Waiting.erase(pos);
 		}
 
-		break;
+		throw;
 	}
 }
 
-CpuBoundWork::~CpuBoundWork()
+class SetAsioCV
 {
-	if (!m_Done) {
-		IoEngine::Get().m_CpuBoundSemaphore.fetch_add(1);
+public:
+	inline SetAsioCV(Shared<AsioConditionVariable>::Ptr cv) : m_CV(std::move(cv))
+	{
 	}
-}
-
-void CpuBoundWork::Done()
-{
-	if (!m_Done) {
-		IoEngine::Get().m_CpuBoundSemaphore.fetch_add(1);
 
-		m_Done = true;
+	inline void operator()()
+	{
+		m_CV->Set();
 	}
-}
 
-IoBoundWorkSlot::IoBoundWorkSlot(boost::asio::yield_context yc)
-	: yc(yc)
-{
-	IoEngine::Get().m_CpuBoundSemaphore.fetch_add(1);
-}
+private:
+	Shared<AsioConditionVariable>::Ptr m_CV;
+};
 
-IoBoundWorkSlot::~IoBoundWorkSlot()
+void CpuBoundWork::Done()
 {
-	auto& ioEngine (IoEngine::Get());
+	if (!m_Done) {
+		auto& sem (IoEngine::Get().m_CpuBoundSemaphore);
+		std::unique_lock<std::mutex> lock (sem.Mutex);
 
-	for (;;) {
-		auto availableSlots (ioEngine.m_CpuBoundSemaphore.fetch_sub(1));
+		if (sem.Waiting.empty()) {
+			++sem.FreeSlots;
+		} else {
+			auto next (sem.Waiting.front());
 
-		if (availableSlots < 1) {
-			ioEngine.m_CpuBoundSemaphore.fetch_add(1);
-			IoEngine::YieldCurrentCoroutine(yc);
-			continue;
+			*next.GotSlot = true;
+			sem.Waiting.pop_front();
+			boost::asio::post(*next.Strand, SetAsioCV(std::move(next.CV)));
```

Comment:
I would just use something like this instead and drop the intermediate class `SetAsioCV` entirely.
```cxx
boost::asio::post(*next.Strand, [cv = std::move(next.CV)]() { cv->Set(); });
```